client/mm/wrapped_cex.go

```diff
@@ -0,0 +1,292 @@
+
+	trade, found := w.trades[update.TradeID]
+	if !found {
+		w.log.Errorf("wrappedCEX: trade ID %s not found", update.TradeID)
+		return
+	}
+
+	if trade.sell && update.QuoteFilled > trade.quoteFilled {
+		quoteFilledDelta := update.QuoteFilled - trade.quoteFilled
+		w.mm.modifyBotCEXBalance(w.botID, trade.toAsset, quoteFilledDelta, balanceModIncrease)
+		trade.quoteFilled = update.QuoteFilled
+		trade.baseFilled = update.BaseFilled
+	}
+
+	if !trade.sell && update.BaseFilled > trade.baseFilled {
+		baseFilledDelta := update.BaseFilled - trade.baseFilled
+		w.mm.modifyBotCEXBalance(w.botID, trade.toAsset, baseFilledDelta, balanceModIncrease)
+		trade.baseFilled = update.BaseFilled
+		trade.quoteFilled = update.QuoteFilled
+	}
+
+	if !update.Complete {
+		return
+	}
+
+	if trade.sell && trade.qty > trade.baseFilled {
+		unfilledQty := trade.qty - trade.baseFilled
+		w.mm.modifyBotCEXBalance(w.botID, trade.fromAsset, unfilledQty, balanceModIncrease)
+	}
+
+	if !trade.sell && calc.BaseToQuote(trade.rate, trade.qty) > trade.quoteFilled {
+		unfilledQty := calc.BaseToQuote(trade.rate, trade.qty) - trade.quoteFilled
+		w.mm.modifyBotCEXBalance(w.botID, trade.fromAsset, unfilledQty, balanceModIncrease)
+	}
+
+	delete(w.trades, update.TradeID)
+}
+
+func (w *wrappedCEX) SubscribeTradeUpdates() (<-chan *libxc.TradeUpdate, func()) {
+	w.subscriptionIDMtx.Lock()
+	defer w.subscriptionIDMtx.Unlock()
+	if w.subscriptionID != nil {
+		w.log.Errorf("SubscribeTradeUpdates called more than once by bot %s", w.botID)
+		return nil, nil
+	}
+
+	updates, unsubscribe, subscriptionID := w.CEX.SubscribeTradeUpdates()
+	w.subscriptionID = &subscriptionID
+
+	ctx, cancel := context.WithCancel(context.Background())
+	forwardUnsubscribe := func() {
+		cancel()
+		unsubscribe()
+	}
+	forwardUpdates := make(chan *libxc.TradeUpdate, 256)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case note := <-updates:
+				w.handleTradeUpdate(note)
+				forwardUpdates <- note
+			}
+		}
+	}()
+
+	return forwardUpdates, forwardUnsubscribe
+}
+
+func (w *wrappedCEX) Trade(ctx context.Context, baseID, quoteID uint32, sell bool, rate, qty uint64) (string, error) {
+	var fromAssetID, toAssetID uint32
+	var fromAssetQty uint64
+	if sell {
+		fromAssetID = baseID
+		toAssetID = quoteID
+		fromAssetQty = qty
+	} else {
+		fromAssetID = quoteID
+		toAssetID = baseID
+		fromAssetQty = calc.BaseToQuote(rate, qty)
+	}
+
+	fromAssetBal := w.mm.botCEXBalance(w.botID, fromAssetID)
+	if fromAssetBal < fromAssetQty {
+		return "", fmt.Errorf("asset bal < required for trade (%d < %d)", fromAssetBal, fromAssetQty)
+	}
+
+	w.mm.modifyBotCEXBalance(w.botID, fromAssetID, fromAssetQty, balanceModDecrease)
+	var success bool
+	defer func() {
+		if !success {
+			w.mm.modifyBotCEXBalance(w.botID, fromAssetID, fromAssetQty, balanceModIncrease)
+		}
+	}()
+
+	w.tradesMtx.Lock()
+	defer w.tradesMtx.Unlock()
+
+	w.subscriptionIDMtx.RLock()
+	defer w.subscriptionIDMtx.RUnlock()
```

Comment:
I think we'd want to avoid nested locks.

Reply:
subscriptionID could be atomic actually.. but what's wrong with locking two things at once?

Reply:
Nothing really wrong since they are protecting different data. But might lead to unexpected issues, I think.

Reply:
Yeah it could cause a deadlock if in another location the locks are acquired in the reverse order.